### ISSUE TITLE
bpftrace not able to attack kprobe with nf_nat with kernel 6.1.55-1

### ISSUE THREAD

**Issue by pkopensrc** [iovisor/bpftrace#2888]
With latest debian kernel 6.1. bpftrace is not able to attach kprobe for nf_nat functions.
It gives error - cannot attach kprobe, probe entry may not exist

### What reproduces the bug? Provide code if possible.
bpftrace -e 'kprobe:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'

/ # bpftrace -e 'kprobe:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'
Attaching 1 probe...
cannot attach kprobe, probe entry may not exist
ERROR: Error attaching probe: 'kprobe:nf_nat_ipv4_manip_pkt'

Mentioned below is working and non working output:
**Not working :**

/ # bpftrace --info
System
  OS: Linux 6.1.0-15-cloud-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.66-1 (2023-12-09)
  Arch: x86_64

Build
  version: v0.17.1
  LLVM: 16.0.3
  unsafe uprobe: no
  bfd: yes
  libdw (DWARF support): yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: no

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes
  kprobe_multi: no
  raw_tp_special: yes

/ # 

$ sudo bpftrace -l | grep "manip"
kfunc:nf_nat:l4proto_manip_pkt
kfunc:nf_nat:nf_nat_ipv4_manip_pkt
kfunc:nf_nat:nf_nat_ipv6_manip_pkt
kfunc:nf_nat:nf_nat_manip_pkt
kprobe:l4proto_manip_pkt
kprobe:nf_nat_ipv4_manip_pkt
kprobe:nf_nat_ipv6_manip_pkt
kprobe:nf_nat_manip_pkt

/ # bpftrace -e 'kprobe:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'
Attaching 1 probe...
cannot attach kprobe, probe entry may not exist
ERROR: Error attaching probe: 'kprobe:nf_nat_ipv4_manip_pkt'

/ # bpftrace -e 'kfunc:nf_nat:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'
Attaching 1 probe...

func called
func called
func called
^C

/ #

**Working:**
/ # bpftrace --info
System
  OS: Linux 6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29)
  Arch: x86_64

Build
  version: v0.17.1
  LLVM: 16.0.3
  unsafe uprobe: no
  bfd: yes
  libdw (DWARF support): yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: no

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes
  kprobe_multi: no
  raw_tp_special: yes

/ # bpftrace -l | grep "manip"
kprobe:l4proto_manip_pkt
kprobe:nf_nat_ipv4_manip_pkt
kprobe:nf_nat_ipv6_manip_pkt
kprobe:nf_nat_manip_pkt
/ #

/ # bpftrace --version
bpftrace v0.17.1
/ # bpftrace -e 'kprobe:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'
Attaching 1 probe...
func called
func called

^C

/ # bpftrace -e 'kfunc:nf_nat:nf_nat_ipv4_manip_pkt { printf("func called\n"); }'
Attaching 1 probe...
func called
func called
func called

^C
/ #

**Comment by jordalgo** [comment]
My first guess is that function was inlined. Can you try running this against both vmlinux binaries?
`llvm-dwarfdump /path/to/vmlinux | grep -A 5 DW_TAG_inlined_subroutine | grep -C 1 nf_nat_ipv4_manip_pkt`

**Comment by pkopensrc** [comment]
As attached It was working fine till 6.1.55 and broken recently.
It it was inline it shouldn't show in bpftrace -l.
Here is output:
root@pk-ntest-12new:/boot# llvm-dwarfdump-14 vmlinuz-6.1.0-15-cloud-amd64 | grep -A 5 DW_TAG_inlined_subroutine | grep -C 1 nf_nat_ipv4_manip_pkt
root@pk-ntest-12new:/boot# 

No changes in function.
static bool nf_nat_ipv4_manip_pkt(struct sk_buff *skb,
				  unsigned int iphdroff,
				  const struct nf_conntrack_tuple *target,
				  enum nf_nat_manip_type maniptype)

**Comment by danobi** [comment]
Hmm, weird. You've already established that it's the change in kernel - not in bpftrace. That is good information to know. But nothing obvious stands out to me when looking at `nf_nat_ipv4_manip_pkt` in master, 6.1, or 6.0. I'm afraid the only way I know to debug this is basically to printk debug the kernel or start bisecting kernels. Neither of which we can really help you with from here.

Bisecting might be easier if you're not familiar with the kernel codepaths in question.

**Comment by jordalgo** [comment]
@pkopensrc Are you able to duplicate this behavior with a raw bcc or libbpf program? Then at least we'd know if it is isolated to bpftrace.

**Comment by pkopensrc** [comment]
I am not a kernel expert and don't know how to  compile kernel and trying my hands in bpf.

As given in  bug output in kernel 6.1.55 it works fine using bpftrace.

I have 2 gcloud VM with different debian and saw issue.

I am not sure its issue with bpftrace or kernel itself. 
Basic libbpf program works fine. 

If its not bpftrace issue How to raise this issue with kernel team? 
I tried to post query to netfilter-dev team but didn't got any respond.

**Comment by jordalgo** [comment]
> Libbpf i tried and it failed

Ah interesting. I wonder if the bug is in bpftrace's listing of kprobes that are not actually available to attach to. I'll see if I can dig in a bit more this weekend.

**Comment by pkopensrc** [comment]
@jordalgo 
Yes that can be also true. If its bug in listing of kprobes. Can you please check that.

But i am getting confused on how it was working in 1.55 and not in 1.66 means something changed recently.
A Kernel developer with an expertise in bpftrace can tell us why its not working.

**Comment by jordalgo** [comment]
@pkopensrc Mind trying one more thing, can you run this on both kernel versions?
`sudo cat /sys/kernel/debug/tracing/available_filter_functions | grep "nf_nat_ipv4_manip_pkt"`

**Comment by pkopensrc** [comment]
@jordalgo here is output which is same in both.

**Comment by jordalgo** [comment]
Well this is definitely a strange one. I'm not able to repro this behavior on ubuntu, centos, fedora, or debian (6.1.0-13-arm64, which is after loading the nf_nat kernel module).

**Comment by pkopensrc** [comment]
Maybe you can try latest kernel 6.1.66 after Dec 12th release

**Comment by jordalgo** [comment]
Ok, I was finally able to repro. Also asked around and it might be that this patch needs to get merged to fix this issue:
https://patchwork.kernel.org/project/netdevbpf/patch/20231027233126.2073148-1-andrii@kernel.org/

**Comment by pkopensrc** [comment]
Thank you for finding out.
but it has to go to main stream kernel. Otherwise cloud providers will not work.

One more strange thing was if you see latest kernel Linux 6.2.0-1019-gcp even all this symbols are not present.

When i asked in dev channel they said kprobe/kfuncs are unreliable and unstable and can change anytime.

**Comment by jordalgo** [comment]
It was just merged into 6.1 stable:
https://lore.kernel.org/stable/20240103164911.126685591@linuxfoundation.org/T/#u

> One more strange thing was if you see latest kernel Linux 6.2.0-1019-gcp even nat symbols are not present in bpftrace.

Did you load the nt_nat module?

> kprobe/kfuncs are unreliable and unstable and can change anytime

This is true and oftentimes why people choose to use tracepoints instead.

**Comment by pkopensrc** [comment]
No how to load same?
Generally in 6.1.55 i don't have to load anything it used to come by default.
